arch/cortex-m/src/dwt.rs

```diff
@@ -0,0 +1,525 @@
+        /// RW.
+        DATAVSIZE   OFFSET(10)   NUMBITS(2),
+
+        /// Reads as one if a second linked comparator is supported.
+        LNK1ENA     OFFSET(9)    NUMBITS(1),
+
+        /// Enables data value comparison
+        /// When 0: Perform address comparison, when 1: data value comparison.
+        /// RW.
+        DATAVMATCH  OFFSET(8)    NUMBITS(1),
+
+        /// Write 1 to enable generation of data trace address packets.
+        /// WARN: If `Control::NOTRCPKT` reads as zero, this bit is UNKNOWN.
+        /// RW.
+        EMITRANGE   OFFSET(5)    NUMBITS(1),
+
+        /// Selects action taken on comparator match.
+        /// Resets to 0b0000.
+        /// RW.
+        FUNCTION    OFFSET(0)    NUMBITS(4),
+    ],
+
+    Comparator3[
+        /// Reference value for comparator 0.
+        COMP       OFFSET(0)   NUMBITS(32),
+    ],
+
+    Comparator3Mask[
+        /// Size of ignore mask aplied to the access address for address range matching by comparator 0.
+        /// WARN: Maximum Mask size is IMPLEMENTATION DEFINED.
+        MASK       OFFSET(0)   NUMBITS(5),
+    ],
+
+    Comparator3Function[
+        /// Is one if comparator matches. Reading the register clears it to 0.
+        /// RO.
+        MATCHED     OFFSET(24)   NUMBITS(1),
+
+        /// Second comparator number for linked address comparison.
+        /// Works, when `DATAVMATCH` and `LNK1ENA` read as one.
+        /// RW.
+        DATAVADDR1  OFFSET(16)   NUMBITS(4),
+
+        /// Comparator number for linked address comparison.
+        /// Works, when `DATAVMATCH` reads as one.
+        /// RW.
+        DATAVADDR0  OFFSET(12)   NUMBITS(4),
+
+        /// Size of data comparision (Byte, Halfword, Word).
+        /// RW.
+        DATAVSIZE   OFFSET(10)   NUMBITS(2),
+
+        /// Reads as one if a second linked comparator is supported.
+        LNK1ENA     OFFSET(9)    NUMBITS(1),
+
+        /// Enables data value comparison
+        /// When 0: Perform address comparison, when 1: data value comparison.
+        /// RW.
+        DATAVMATCH  OFFSET(8)    NUMBITS(1),
+
+        /// Write 1 to enable generation of data trace address packets.
+        /// WARN: If `Control::NOTRCPKT` reads as zero, this bit is UNKNOWN.
+        /// RW.
+        EMITRANGE   OFFSET(5)    NUMBITS(1),
+
+        /// Selects action taken on comparator match.
+        /// Resets to 0b0000.
+        /// RW.
+        FUNCTION    OFFSET(0)    NUMBITS(4),
+    ],
+
+];
+
+const DWT: StaticRef<DwtRegisters> = unsafe { StaticRef::new(0xE0001000 as *const DwtRegisters) };
+
+struct Dwt {
+    registers: StaticRef<DwtRegisters>,
+}
+/// Returns wether a cycle counter is present on the chip.
+pub fn is_cycle_counter_present() -> bool {
+    if DWT.ctrl.read(Control::NOCYCCNT) == 0 {
+        true
+    } else {
+        false
+    }
+}
+
+impl Dwt {
+    pub const fn new() -> Self {
+        Self { registers: DWT }
+    }
+}
+
+impl hil::debug::PerformanceCounters for Dwt {
```

Comment:
I think it would also be helpful to add a `reset_cycle_counter()` method to this HIL to reset the counter to 0. This saves from having to get the current count and then perform subtraction, and it just an easy interface (imo).  Amit wrote a rough DWT implementation a few years ago that I often use when benchmarking, and it just does the following:

```rust
pub fn reset_timer() {
    DEMCR.set(DEMCR.get() | 0x01000000);
    DWT.cycnt.set(0); // reset the counter
    DWT.ctrl.set(0); // disable counter;
}
```

Keeping in mind that on something like the nrf52 this counter will overflow in 67 seconds, it can actually be pretty helpful to be able to reset to 0 rather than have to consider the (somewhat high) possibility of overflow.